File: Linked_List/singleLinkedList.go
package main

import "fmt"

type Node struct {
	val  int
	next *Node
}
type LinkedList struct {
	head *Node
	tail *Node
	size int
}

func (L *LinkedList) isEmpty() bool {
	return L.size == 0
}
func (L *LinkedList) display() {
	p := L.head
	for p != nil {
		fmt.Printf("%d ->", p.val)
		p = p.next
	}
	fmt.Println("")
}
func (L *LinkedList) addLast(val int) {
	newest := &Node{val, nil}
	if L.isEmpty() {
		L.head = newest
	} else {
		L.tail.next = newest
	}
	L.tail = newest
	L.size++
}
func (L *LinkedList) addFirst(val int) {
	newest := &Node{val, nil}
	if L.isEmpty() {
		L.head = newest
		L.tail = newest
	} else {
		newest.next = L.head
		L.head = newest
	}
	L.size++

}
func (L *LinkedList) addany(e, pos int) {
	newest := &Node{e, nil}
	p := L.head
	i := 1
	for i < pos-1 {
		p = p.next
		i++
	}
	newest.next = p.next
	p.next = newest
	L.size++

}
func (L LinkedList) search(key int) bool {
	for L.head != nil {
		if L.head.val == key {
			return true
		}
		L.head = L.head.next

	}
	return false
}
func (L *LinkedList) removeLast() int {
	if L.isEmpty() {
		fmt.Println("list is empty")
		return -1
	}
	i := 1
	p := L.head
	for i < L.size-1 {
		p = p.next
		i++
	}
	e := L.tail.val
	L.tail = p
	p.next = nil
	L.size--

	if L.size == 0 {
		L.tail = nil
	}
	return e
}
func (L *LinkedList) removeFirst() int {
	if L.isEmpty() {
		fmt.Println("list is empty")
		return -1
	}
	e := L.head.val
	L.head = L.head.next
	L.size--
	return e
}
func (L *LinkedList) removeany(pos int) {
	p := L.head
	i := 1
	for i < pos-1 {
		p = p.next
		i++
	}
	p.next = p.next.next
	L.size--

}

func main() {
	fmt.Println("goSingleLinkedList")
	n := LinkedList{}
	fmt.Println(n)
	fmt.Println(n.size)
	n.addLast(34)

	fmt.Println(n.size)

	n.display()
	n.addLast(20)

	fmt.Println(n.size)

	n.display()
	n.addLast(50)

	fmt.Println(n.size)
	n.display()
	n.addFirst(5)
	n.addFirst(1)
	fmt.Println(n.size)
	n.display()
	fmt.Println(n.search(-20))
	fmt.Println(n.removeLast())
	fmt.Println(n.size)
	n.display()
	fmt.Println(n.removeFirst())
	fmt.Println(n.size)
	n.display()
	n.addany(25, 2)
	fmt.Println(n.size)
	n.display()
	n.removeany(1)
	fmt.Println(n.size)
	n.display()
	fmt.Println(n.removeFirst())
	fmt.Println(n.removeFirst())
	fmt.Println(n.size)
	n.display()
	n.removeFirst()
	fmt.Println(n.size)
	n.display()

}

File: Linked_List/single_linked_list.py
class  _Node :
    __slots__ = "_element","_next"
    def __init__(self,element,next):
        self._element=element 
        self._next=next 
class Linked_List:
    def __init__(self):
        self.head=None 
        self.tail=None 
        self.size=0
    def __len__(self):
        return self.size 
    def isempty(self):
        return self.size==0 
    def addlast(self, e):
        newest=_Node(e,None)
        if self.isempty():
            self.head=newest
        else:
            self.tail._next=newest
        
        self.tail=newest 
        self.size+=1
    def addfirst(self, e):
        newest=_Node(e,None) 
        if self.isempty():
            self.head=newest 
            self.tail=newest 
        else:
            newest._next=self.head 
            self.head=newest 
        self.size+=1 
    def addany(self,e,position):
        newest=_Node(e,None)
        p=self.head 
        i=1 
        while i<position-1:
            p=p._next 
            i+=1 

        newest._next=p._next 
        p._next=newest 
        self.size+=1 
    def removefist(self):
        if self.isempty():
            print("List is empty")
            return 
        e=self.head._element 
        self.head=self.head._next 
        self.size-=1 
        if self.isempty():
            self.tail=None 
        return e 
    def removelast(self):
        if self.isempty():
            print("List is empty")
            return 
        p=self.head 
        i=1 
        while i<len(self)-1:
            p=p._next 
            i+=1 
        self.tail=p 
        e=p._next._element 
        self.tail._next=None 
        self.size-=1 
        return e

    def removeany(self,position):
        if self.isempty():
            print("List is empty")
            return 
        p=self.head 
        i=1 
        while i<position-1:
            p=p._next 
            i+=1 
        e=p._next._element 
        p._next=p._next._next 
        self.size-=1 
        return e 
    def display(self):
        p=self.head 
        while p:
            print(p._element, end="--->")
            p=p._next 
        print()
    def search(self,key):
        p=self.head 
        index=0 
        while p:
            if p._element==key:
                return index 
            p=p.next 
            index+=1 
        return -1 

L = Linked_List()
L.addfirst(4)
L.addlast(7)
L.addlast(4)
L.addlast(12)
L.addlast(8)
L.addlast(3)
L.display()
print('Size:',len(L))
ele = L.removeany(3)
L.display()
print('Size:',len(L))
print('Removed Element:',ele)

    







